### ISSUE TITLE
How to show a modal dialog from a Host Object?

### ISSUE THREAD

**Issue by cylvoon** [MicrosoftEdge/WebView2Feedback#2300]
Hi,

I'm migrating some existing code from CHtmlView to WebView2 in a C++/MFC project.
We used to have an External that showed a modal dialog and return a result to the htmlview depending on a choice in that dialog.
With WebView2, I created a Host Object and added it using AddHostObjectToScript.
My Host Object gets called but if I try to show a modal dialog, it breaks the debugger with the following message: "Running a message loop synchronously in an event handler in Webview can cause reentrancy issue".

From what I read here (https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/threading-model#re-entrancy), Webview2 design prevents from starting a message loop in an event handler.
Ok, but what's the workaround?

On a side note, I hate it that WebView2 team has completely ignored old CHtmlView.
It's a real mess to migrate from to the other.
And we don't have the choice as IE is not maintained and disappearing soon anyway.

**Comment by cylvoon** [comment]
@champnic Would you mind looking at this?

**Comment by champnic** [comment]
Hey @cylvoon - thanks for the question. In general, you just need to make sure you aren't within the context of the event handler. In the docs you can see they workaround this in .NET by using a dispatcher. In C++ I think the equivalent would be to post a message to show the dialog, so that it's not happening from within the event handler. @LiangTheDev is that right?

re; CHtmlView - the IE deprecation is only for the browser. The web controls built on IE will continue to work.

**Comment by LiangTheDev** [comment]
champnic's comment above is right. Some comments though:

- Moving the dialog out of the host object callback/event handler would mean that you have to modify the code to send the result back to the web page, which means more changes.
- Showing a modal dialog actually could mostly work if we are not running under debugger. It just means that we will not be able to interact with any WebView2 object or process any events/messages from them while the dialog is shown. If you are not doing anything related to WebView2 when the dialog is shown, you can use the code as it is for now.

**Comment by cylvoon** [comment]
@champnic Thanks for the reply. This is the kind of solution I was heading to, however, as @LiangTheDev pointed out, it requires a lot of changes on c++ and on JS side.
About IE, you mean that until further notice, CHtmlView should continue to work even on windows without IE installed (like Win11)?

@LiangTheDev Thanks for your input. You're definitely right about the amount of changes required.
But it looks like I have no choice, despite the fact I had not realized it could work outside of the debugger, I have some dialog (triggered by w webview2 event) that embed their own webview2. In that case, without debugger, those dialogs stay blank (which makes sense, unfortunately).

**Comment by cylvoon** [comment]
@champnic One detail, as I wrote earlier "We used to have an External that showed a modal dialog and return a result to the htmlview depending on a choice in that dialog.". This means I cannot just post a message to show my dialog as you suggested, as I need a synchronous result. So I have to make all the existing code asynchronous which is, again, a lot of work!
Really a bummer you guys did not think of that.
